keps/sig-auth/20191204-serviceaccount-certificate.md

```diff
@@ -0,0 +1,373 @@
+
+With the ServiceAccount x509 certificates implemented, 
+Ping and Pong could use ServiceAccount x509 certificate as their ServerAuth and ClientAuth.
+
+For example, the Ping Pod may like that:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: ping
+spec:
+  serviceAccountName: ping-sa
+  containes:
+  - name: ping
+    image: ping:latest
+    volumeMounts:
+    - name: tls-client-k8s
+      path: /var/run/secrets/kubernetes.io/serviceaccount
+    - name: tls-client-pong
+      path: /tls-client-pong
+    - name: tls-server
+      path: /tls-cerver
+  volumes:
+    - name: tls-client-pong
+      projected:
+        configMap:
+          name: kube-root-ca.crt
+          items:
+            - key: ca.crt
+              path: ca.crt
+        serviceAccountTLS: 
+          usage:
+          - ClientAuth
+          extensions:
+          - "client-name=ping"
+          certificatePath: tls.crt
+          certificateKey: tls.key
+    - name: tls-server
+      projected:
+        serviceAccountTLS: 
+          usage:
+          - ServerAuth
+          serveHosts:
+          - ping
+          - ping.default
+          - ping.default.svc
+          certificatePath: tls.crt
+          certificateKey: tls.key
+  ...
+```
+
+There are 3 ServiceAccount x509 certificates exist in the Ping Pod:
+
+1. *tls-client-k8s*: mount in `/var/run/secrets/kubernetes.io/serviceaccount`, 
+which Ping used as ClientAuth to visit Kubernetes, Kubernetes could get ServiceAccount namespace and name from Ping's request.
+2. *tls-client-pong*: mount in `/tls-client-pong`, which Ping used as ClientAuth to visit Pong Service. 
+The Pong could get these information from Ping's request certificate and do Authz then:
+   1. Client ServiceAccount namespace and name.
+   2. Extensions, which is `client-name=ping` in this example.
+3. *tls-server*: mount in `/tls-cerver` which Ping used as its HTTPs TLS certificate. 
+When Pong requests to Ping, Pong will check Ping's TLS certificate is trusted.
+
+The Pong example omitted, as this just like a mirror of Ping.
+
+### Risks and Mitigations
+
+TBD, let's talk about ServiceAccount x509 certificate rotate and CA/CA-Key leak.
+
+## Design Details
+
+### API Change
+
+#### Extend ServiceAccountTLSProjection to VolumeProjection
+
+`ServiceAccountTLSProjection` is used by a Pod to request and mount the ServiceAccount x509 certificates:
+
+```go
+// ServiceAccountTLSProjection represents a projected service account x509 certificate
+// volume. This projection can be used to insert a service account certificate into
+// the pods runtime filesystem for use against APIs (Kubernetes API Server or
+// otherwise).
+type ServiceAccountTLSProjection struct {
+  // Usages is the usage of TLS certificate.
+  Usages []ServiceAccountTLSUsage
+  
+  // ServeHosts is the SANs for DNS or IPs.
```

Comment:
Would URI support be desired? This could allow usage of SPIFFE certificates and potentially use in service meshes that rely on these